dig/dig.go

```diff
@@ -0,0 +1,321 @@
+	Reset()
+}
+
+// Register an object in the dependency graph
+//
+// Provided argument must be a function that returns exactly one pointer argument
+// All arguments to the function must be pointers
+func (g *graph) Register(c interface{}) error {
+	g.Lock()
+	defer g.Unlock()
+
+	ctype := reflect.TypeOf(c)
+
+	switch ctype.Kind() {
+	case reflect.Func:
+		if ctype.NumOut() != 1 {
+			return errReturnCount
+		}
+
+		objType := ctype.Out(0)
+		if objType.Kind() != reflect.Ptr && objType.Kind() != reflect.Interface {
+			return errReturnKind
+		}
+
+		return g.registerConstructor(c)
+	case reflect.Ptr:
+		return g.registerObject(c)
+	default:
+		return errParamType
+	}
+}
+
+// Resolve all of the dependencies of the provided class
+//
+// Provided object must be a pointer
+// Any dependencies of the object will receive constructor calls, or be initialized (once)
+// Constructor with return value *object will be called
+//
+// TODO(glib): catch any and all panics from this method, as there is a lot of reflect going on
+func (g *graph) Resolve(obj interface{}) error {
+	g.Lock()
+	defer g.Unlock()
+
+	objType := reflect.TypeOf(obj)
+	if objType.Kind() != reflect.Ptr {
+		return fmt.Errorf("can not resolve non-pointer object of type %v", objType)
+	}
+
+	objElemType := reflect.TypeOf(obj).Elem()
+	objVal := reflect.ValueOf(obj)
+
+	// check if the type is a registered objNode
+	n, ok := g.nodes[objElemType]
+	if !ok {
+		return fmt.Errorf("type %v is not registered", objType)
+	}
+
+	v, err := n.value(g)
+	if err != nil {
+		return errors.Wrapf(err, "unable to resolve %v", objType)
+	}
+
+	// set the pointer value of the provided object to the instance pointer
+	objVal.Elem().Set(v)
+
+	return nil
+}
+
+// ResolveAll the dependencies of each provided object
+// Returns the first error encountered
+func (g *graph) ResolveAll(objs ...interface{}) error {
+	for _, o := range objs {
+		if err := g.Resolve(o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RegisterAll registers all the provided args in the dependency graph
+func (g *graph) RegisterAll(cs ...interface{}) error {
+	for _, c := range cs {
+		if err := g.Register(c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Reset the graph by removing all the registered nodes
+func (g *graph) Reset() {
+	g.Lock()
+	defer g.Unlock()
+
+	defaultGraph.nodes = make(map[interface{}]object)
+}
+
+type object interface {
+	// Return value of the object
+	value(g *graph) (reflect.Value, error)
+
+	// Other things that need to be present before this object can be created
+	dependencies() []interface{}
+}
+
+type graph struct {
+	fmt.Stringer
+	sync.Mutex
+
+	nodes map[interface{}]object
+}
+
+func newGraph() *graph {
+	return &graph{
+		nodes: make(map[interface{}]object),
+	}
+}
+
+func (g *graph) String() string {
+	b := &bytes.Buffer{}
+	fmt.Fprintln(b, "{nodes:")
+	for key, reg := range g.nodes {
+		fmt.Fprintln(b, key, " -> ", reg)
```

Comment:
Good idea. Im not sure that preserving the order of dependencies from the constructor in the right order is that important for debugging, but if it gets to that I will extend it as such.